proposals/lip-0053.md

```diff
@@ -34,127 +34,262 @@ Identifies the chain sending the cross-chain update. Only cross-chain update tra
+```python
+def verifyValidatorsUpdate(ccu: CCU) -> None:
+    if len(ccu.params.certificate) == 0:
+        raise Exception("Certificate must be non-empty if validators have been updated.")
+    
+    partnerchainID = ccu.params.sendingChainID
+    validatorKeys = [validator.blsKey for validator in ccu.params.activeValidatorsUpdate]
+    # All validator keys must be distinct.
+    if len(validatorKeys) != len(set(validatorKeys)):
+        raise Exception("Duplicate BLS keys.")
+    
+    # Validator keys must be in lexicographic order.
+    if not all(validatorKeys[i] < validatorKeys[i + 1] for i in range(len(validatorKeys) - 1)):
+        raise Exception("Validator keys are not in lexicographic order.")
+
+    # Update validator list.
+    newActiveValidators = calculateNewActiveValidators(validators(partnerchainID).activeValidators, ccu.params.activeValidatorsUpdate)
+    newCertificateThreshold = ccu.params.newCertificateThreshold
+    # computeValidatorsHash is defined in LIP 0058.
+    certificate = decode(certificateSchema, ccu.params.certificate)
+    if certificate.validatorsHash == bft.computeValidatorsHash(newActiveValidators, newCertificateThreshold):
+        raise Exception("Validators hash does not match.")
+```
 
-## Specification
+#### calculateNewActiveValidators
+
+```python
+def calculateNewActiveValidators(activeValidators: list[ActiveValidator], activeValidatorsUpdate: list[ActiveValidator]) -> list[ActiveValidator]:
+    validatorKeys = [validator.blsKey for validator in activeValidators]
+    for validator in activeValidatorsUpdate:
+        for currentValidator in activeValidators:
+            if currentValidator.blsKey == validator.blsKey:
+                currentValidator.bftWeight = validator.bftWeight
+        else:
+            add validator to activeValidators maintaining the entries in lexicographical order of blsKeys
 
-The Interoperability module supports two commands used to certify the state of another chain. Those commands have `moduleID = MODULE_ID_INTEROPERABILITY`. One of them, meant to be posted on the mainchain, has `commandID = COMMAND_ID_MAINCHAIN_CCU`, while the other, meant to be posted on sidechains, has `commandID = COMMAND_ID_SIDECHAIN_CCU`.
+    #remove any entry activeValidators with bftWeight == 0
+    for validator in activeValidators:
+        if validator.bftWeight == 0:
+            activeValidators.remove(validator)
+    return activeValidators
+```
 
-### Constants and Notations
+#### verifyPartnerChainOutboxRoot
 
-The following constants are used throughout the document:
+```python
+def verifyPartnerChainOutboxRoot(ccu: CCU) -> None:
+    partnerchainID = ccu.params.sendingChainID
+    inboxTree = Merkle tree of channel(partnerchainID).inbox
+    # Update inbox by appending all cross-chain messages.
+    # crossChainMessages can be empty.
+    for ccmBytes in ccu.params.inboxUpdate.crossChainMessages:
+        inboxTree.append(sha256(ccmBytes))
+    newInboxRoot = inboxTree.root
```

Comment:
Should we remove this line as this variable gets overwritten in any case below?

Reply:
removed